rust/index/src/spann/utils.rs

```diff
@@ -0,0 +1,565 @@
+use std::{cmp::min, collections::HashMap};
+
+use chroma_distance::DistanceFunction;
+use rand::{seq::SliceRandom, Rng};
+
+// TODO(Sanket): I don't understand why the reference implementation defined
+// max_distance this way.
+const MAX_DISTANCE: f32 = f32::MAX / 10.0;
+const NUM_ITERS_FOR_CENTER_INIT: usize = 3;
+const NUM_ITERS_FOR_MAIN_LOOP: usize = 100;
+const NUM_ITERS_NO_IMPROVEMENT: usize = 5;
+
+struct KMeansAlgorithmInput<'referred_data> {
+    indices: Vec<u32>,
+    embeddings: &'referred_data [f32],
```

Comment:
nit - might be nice to have a wrapper over flat embedding / embedding dimension

Reply:
need to look at this wholistically since there are a few other places also that have flat embeddings. Cut out a task in tracker for this.